### PR TITLE
Add support for versioned BokehJS script tags and CDN urls

### PR DESCRIPTION
**Pull request by santos22**
- [x] issues: fixes https://github.com/bokeh/bokeh/issues/9919

Failed PR: https://github.com/bokeh/bokeh/pull/10075

### CODE REVIEW

**Comment by santos22** [comment]
Will change the pattern matching to match these urls as well:
```javascript
https://cdn.bokeh.org/bokeh/release/bokeh-a7ee96e.min.js
https://cdn.bokeh.org/bokeh/release/bokeh-widgets-a7ee96e.min.js
https://cdn.bokeh.org/bokeh/release/bokeh-tables-a7ee96e.min.js
https://cdn.bokeh.org/bokeh/release/bokeh-gl-a7ee96e.min.js
https://cdn.bokeh.org/bokeh/release/bokeh-api-a7ee96e.min.js
```

File: bokeh/sphinxext/bokehjs_link.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        bokeh_url_pattern = re.compile(rf'^.*(?!{group})')
+    else:
+        bokeh_url_pattern = re.compile(rf'.*{text}.*')
+
+    matched_urls = list(filter(bokeh_url_pattern.match, js_files))
+
+    if not matched_urls:
+        raise SphinxError("cannot find CDN url for %r, urls: %s" % (text, str(js_files)))
+
+    url = matched_urls[0]
```
Comment (santos22) [inline bokeh/sphinxext/bokehjs_link.py:102]: Added error handling and fixed up the pattern matching.

**Comment by bryevdv** [comment]
@santos22 In the users guide this seems to be duplicating the API bundle:

In a "dev" configuration:

Screen Shot 2020-05-31 at 6 42 41 PM

in a "release" configuration:

Screen Shot 2020-05-31 at 6 47 58 PM

Alos looking at how dense that output is now I think it might be good to put a blank line in between each script tag.

**Comment by bryevdv** [comment]
@santos22 more information, specifically I see that when I build like:
```
BOKEH_DOCS_CDN=2.0.2 BOKEH_DOCS_VERSION="2.0.2"  GOOGLE_API_KEY="junk" make clean html serve
```
But the actual release build will also build like that so we need to make sure it works with those env vars set

**Comment by santos22** [comment]
Fixed the empty line between script tags and am still investigating why the api bundle is being duplicated.

This issue is strange because the duplication happens only with the `bokehjs-script-tag` Sphinx directive when calling `get_sphinx_resources` and not with the `bokehjs-link` Sphinx role.

Also, I think the api bundle duplication _might_ be happening in the changes that fixed https://github.com/bokeh/bokeh/pull/10025 😐

**Comment by santos22** [comment]
Still no fix for the duplicate bundles. The following works but doesn't seem like the right approach:
```python
html_source = BJS_SCRIPT_TAGS.render(
    js_files=list(set(resources.js_files)),
    hashes=resources.hashes)
```

**Comment by santos22** [comment]
May have found the root issue by via the Python debugger. It looks like the `resources` variable is being shared across the Sphinx extensions...

The `bokehjs-link` role is first to be processed when building the docs and `bokeh-api` gets appended to the js components list each time (see screenshot below) this role is used. I guess this issue was kind of hidden here since we only take the first result from the matching results.
```rst
* :bokehjs-link:`bokeh`
* :bokehjs-link:`bokeh-widgets`
* :bokehjs-link:`bokeh-tables`
* :bokehjs-link:`bokeh-api`
```

Screen Shot 2020-06-14 at 12 01 39 PM

The duplicate api bundle state then carries over when calling the `bokehjs-script-tag` directive - seems like removing the `bokehjs-link` allows this directive to start off in the right state:
Screen Shot 2020-06-14 at 12 06 57 PM

Will continue investigating

File: bokeh/sphinxext/util.py [hunk 2]
```diff
@@ -59,9 +70,68 @@ def get_sphinx_resources(include_bokehjs_api=False):
         # Otherwise assume it is a dev/rc/full release version and use CDN for it
         else:
             resources = Resources(mode="cdn", version=docs_cdn)
+    sphinx_resources = deepcopy(resources)
```
Comment (santos22) [inline bokeh/sphinxext/util.py:73]: Added `deepcopy()` here to resolve the api bundle duplication issue 👍

File: sphinx/source/docs/user_guide/embed.rst [hunk 3]
```diff
@@ -197,64 +197,34 @@ script, when executed, will replace the div with the plot.
...
     In [1]: import bokeh.resources
 
-    In [2]: bokeh.resources.get_sri_hashes_for_version("2.0.0")
-    Out[2]:
-    {'bokeh-2.0.0.js': 'TQAjsk2/lDn1NHjYoe8HIascd3/Cw4EWdk6GNtYXVVyAiUkbEZiuP7fEgbSwM37Y',
-
-    ...
-
-    'bokeh-widgets-2.0.0.min.js': '2ltAd1cQhavmLeBEZXGgnna8fjbw+FjvDq9m2dig4+8KVS8JcYFUQaALvLT//qHE'}
+    In [2]: bokeh.resources.CDN.render_js()
```
Comment (santos22) [inline sphinx/source/docs/user_guide/embed.rst:221]: The exact output of this is:
```html

<script type="text/javascript" src="https://cdn.bokeh.org/bokeh/release/bokeh-2.0.2.min.js" integrity="sha384-ufR9RFnRs6lniiaFvtJziE0YeidtAgBRH6ux2oUItHw5WTvE1zuk9uzhUU/FJXDp" crossorigin="anonymous"></script>
<script type="text/javascript" src="https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.0.2.min.js" integrity="sha384-8QM/PGWBT+IssZuRcDcjzwIh1mkOmJSoNMmyYDZbCfXJg3Ap1lEvdVgFuSAwhb/J" crossorigin="anonymous"></script>
<script type="text/javascript" src="https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.0.2.min.js" integrity="sha384-Jm8cH3Rg0P6UeZhVY5cLy1WzKajUT9KImCY+76hEqrcJt59/d8GPvFHjCkYgnSIn" crossorigin="anonymous"></script>
<script type="text/javascript" src="https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.0.2.min.js" integrity="sha384-Ozhzj+SI7ywm74aOI/UajcWz+C0NjsPunEVyVIrxzYkB+jA+2tUw8x5xJCbVtK5I" crossorigin="anonymous"></script>
<script type="text/javascript">
    Bokeh.set_log_level("info");
</script>
```

**Comment by bryevdv** [comment]
wondering if this is tangentially related to #10166 

Anyway thanks @santos22 I will take a closer look in the next few days.

**Comment by bryevdv** [comment]
We are going to move the `master` branch to `main` which will automatically close this PR. Since it also has conflicts, and would need to be rebased against a new devlopment branch regardless, I will go ahead and close it now manually.

**Pull request closed without merging**